### PR TITLE
Factor out common components of demo-runtime and merge into polkadot

### PR DESCRIPTION
**Pull request by gavofyork**
#100

### CODE REVIEW

File: substrate/runtime/timestamp/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	fn timestamp_works() {
+
+		let mut t: TestExternalities = map![
+			twox_128(<Now<TraitImpl>>::key()).to_vec() => vec![].and(&42u64)
+		];
+
+		with_externalities(&mut t, || {
+			assert_eq!(<Now<TraitImpl>>::get(), 42);
+//			assert_eq!(Timestamp::NOW.get(), 42);
+			<public::Callable<TraitImpl>>::set(69).dispatch(&0);
```
Comment (gavofyork) [inline substrate/runtime/timestamp/src/lib.rs]: This doesn't work :(

File: substrate/runtime/timestamp/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#[cfg(test)] extern crate substrate_runtime_io as runtime_io;
+extern crate substrate_codec as codec;
+
+#[cfg(feature = "std")] #[macro_use] extern crate serde_derive;
+#[cfg(feature = "std")] extern crate serde;
+
+use runtime_support::storage::StorageValue;
+use runtime_support::PublicPass;
+
+pub trait Trait {
```
Comment (rphmeier) [inline substrate/runtime/timestamp/src/lib.rs]: a bit weird to call it `Trait`

File: substrate/runtime/system/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+	RandomSeed get(random_seed): b"sys:rnd" => required Hash;
+	// The current block number being processed. Set by `execute_block`.
+	Number get(block_number): b"sys:num" => required BlockNumber;
+	ParentHash get(parent_hash): b"sys:pha" => required Hash;
+	TransactionsRoot get(transactions_root): b"sys:txr" => required Hash;
+	Digest: b"sys:dig" => default block::Digest;
+}
+
+pub trait Checkable {
+	type CheckedType;
```
Comment (rphmeier) [inline substrate/runtime/system/src/lib.rs]: docs? hard to follow what this is about. `type Checked` and `trait Check` would be more idiomatic

File: substrate/runtime/system/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	fn check(self) -> Option<Self::CheckedType>;
+}
+
+pub trait Dispatchable {
+	type AccountIdType;
+	type TxOrderType;
```
Comment (rphmeier) [inline substrate/runtime/system/src/lib.rs]: likewise here -- associated type identifiers don't need to be suffixed with "Type"

File: substrate/runtime/timestamp/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	type Timestamp: codec::Slicable;
+	type PublicAux;
+	type PrivAux;
+}
+pub struct Module<T: Trait>(::std::marker::PhantomData<T>);
```
Comment (rphmeier) [inline substrate/runtime/timestamp/src/lib.rs]: if this struct adds nothing to the trait definition on its own, why do we need the `Module` struct definition in the first place?

File: substrate/runtime/timestamp/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+#[cfg(feature = "std")] #[macro_use] extern crate serde_derive;
+#[cfg(feature = "std")] extern crate serde;
+
+use runtime_support::storage::StorageValue;
+use runtime_support::PublicPass;
+
+pub trait Trait {
+	type Timestamp: codec::Slicable;
+	type PublicAux;
+	type PrivAux;
```
Comment (rphmeier) [inline substrate/runtime/timestamp/src/lib.rs]: it seems to me that the only thing which is actually variable here is the timestamp type definition and that the others should follow, while the definition of these `PublicAux` and `PrivAux` as associated types implies that it's expected for them to vary while the timestamp type holds.

File: substrate/runtime/timestamp/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+pub trait StorageItems {
+	type NowS;
+}
+impl<T: Trait> StorageItems for Module<T> {
+	type NowS = Now<T>;
+}
+
+impl<T: Trait> Module<T> {
+	pub fn get() -> T::Timestamp { <Now<T>>::get() }
```
Comment (rphmeier) [inline substrate/runtime/timestamp/src/lib.rs]: and if the timestamp were the only generic param here (and we went a little more general than "some serializable type" to get to "some timestamp which can be get and set") the implementation here would still be the same, while still extending to timestamps which are out of storage _and_ being testable without setting up storage context

File: substrate/runtime/timestamp/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+/*
+impl_dispatch! {
+	trait Trait;
+	pub mod public;
+	fn set(_, now: Timestamp) = 0;
+}
+*/
+
+pub trait Dispatchable {
+	type AuxType;
```
Comment (rphmeier) [inline substrate/runtime/timestamp/src/lib.rs]: same typename grumble
